tools/wake-format/emitter.cpp

```diff
@@ -0,0 +1,217 @@
+/*
+ * Copyright 2022 SiFive, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You should have received a copy of LICENSE.Apache2 along with
+ * this software. If not, you may obtain a copy at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Open Group Base Specifications Issue 7
+#define _XOPEN_SOURCE 700
+#define _POSIX_C_SOURCE 200809L
+
+#include "emitter.h"
+
+#include <cassert>
+
+#include "parser/parser.h"
+
+wcl::rope Emitter::newline(ctx_t ctx) {
+  wcl::rope_builder builder;
+
+  builder.append("⏎\n");
+  for (int i = 0; i < ctx.nest_level; i++) {
+    builder.append(space(space_per_indent));
+  }
+
+  return std::move(builder).build();
+}
+
+wcl::rope Emitter::space(uint8_t count) {
+  wcl::rope_builder builder;
+  for (uint8_t i = 0; i < count; i++) {
+    builder.append("·");
+  }
+  return std::move(builder).build();
+}
+
+wcl::optional<wcl::rope> Emitter::flat(ctx_t ctx, CSTElement node) {
+  return walk_node(ctx.flat(), node);
+}
+
+wcl::rope Emitter::try_flat(ctx_t ctx, CSTElement node) {
+  auto flat_node = flat(ctx, node);
+  if (flat_node && flat_node->column() <= max_column_width) {
```

Comment:
This isn't the correct condition. I don't think `try_flat` is ever possible since you need to know the context its being inserted into. I guess this can curtail the search space somewhat but the maximum search space isn't that big once we memoize.

Reply:
This function is a bit rewritten in the next PR (though still wrong if this is wrong) can we address there?

Reply:
Yeah this is approved. I'm confident that the fix to this bug is nothing catastrophic which is all I was concerned about with the column stuff in the first place. This PR is about getting the API right.